pkg/arvo/neo/cod/std/src/fil/diary.md

```diff
@@ -0,0 +1,297 @@
+
+## +kids
+Every shrub is a node in one tree, with one “root shrub” at the top of that tree. Every shrub below the root shrub is either one of its immediate children or one of its descendants.
+
+In its `+kids` arm, every shrub can define constraints for the shrubs below it in the namespace, whether that’s constraining the types of their state or the pokes they’ll accept from other shrubs.
+
+Let’s expand on diary’s `+kids` arm.
+
+```hoon
+::
+::  constrain shrubs below diary in the namespace
+::  by defining the types of their state and pokes
+++  kids
+  ::
+  ::  kids:neo is a (unit port:neo)
+  ^-  kids:neo
+  %-  some
+  ::
+  ::  port:neo is (pair dare:neo lads:neo)
+  ::  dare:neo is ?(%y %z)
+  ::  if %y, only constrain our immediate children
+  ::  if %z, recursively constrain all descendants
+  :-  %y
+  ::  lads:neo is (map pish:neo lash:neo)
+  %-  ~(gas by *lads:neo)
+  :~  :-  ::
+          ::  pish:neo
+          ::  to simplify: [%.n @da] means the kid's
+          ::  path contains any @da, and %.n is there
+          ::  to signify that the pith can not have more
+          ::  fields afterwards
+          [[%.n %da] %.n]
+      ::
+      ::  lash:neo is (pair curb:neo (set stud:neo))
+      ::  curb:neo defines the kids' state
+      ::  (set stud:neo) defines the kids' pokes
+      [[%only %txt] ~]
+  ==
+```
+
+In this case, Diary only constrains its immediate children. Each of its children lives at `/path/to/diary/<@da>`, where `<@da>` is used as an ID for the entry, and those children are just `%txt`s (cords) which store the text of the diary entry.
+
+There’s no need for Diary to recusrively constrain every single path that nests under `/path/to/diary`, since the app only needs to reserve the first “generation” of shrubs below it in order to work. In this case, the developer has forbidden any shrubs being created under the diary entries by ending the `pish` in a `%.n`. If that were `%.y`, any app could make a shrub at `/path/to/diary/<@da>/comments`. or `/path/to/diary/<@da>/backlinks`.
+
+To define the shrub’s kids, we create an empty map which is the bunt of `lads:neo`, then use `+gas:by` to populate the map with our own data, which will be a `(list (pair pish:neo lash:neo))`.
+
+The `pish:neo` statically types the paths which we’ll allow to be created beneath this shrub, and the `lash:neo` defines the kids’ state and pokes. In this case, the kids take no pokes and their state can only be `%txt`.
+
+(It’s worth flagging that `$curb:neo` contains several combinatorial rules about state types. For example `[%pro %txt]` would mean “the state of the child shrub can be any type which is readily convertible into a `%txt` with a conversion we have available in the `/con` folder”. This gives us a clue as to how we could handle state transitions and interoperability over the lifetime of our shrub, but it’s outside the remit of this tutorial.)
```

Comment:
I thought about a "grand unified theory of `$conf:neo`" section in the Messenger tutorial; if illustrated well it's a selling point and if not illustrated it can look like a dangly bit, but I think we'd need to kick around use-cases for the combinatorial rules.